File: src/boink/dbg.cc
/**
 * (c) Camille Scott, 2019
 * File   : dbg.cc
 * License: MIT
 * Author : Camille Scott <camille.scott.w@gmail.com>
 * Date   : 30.08.2019
 */

#include "boink/dbg.hh"

#include "boink/storage/storage_types.hh"
#include "boink/hashing/shifter_types.hh"


namespace boink {

    template class dBG<storage::BitStorage, hashing::FwdLemireShifter>;
    template class dBG<storage::BitStorage, hashing::CanLemireShifter>;
    template class dBG<storage::BitStorage, hashing::FwdUnikmerShifter>;
    template class dBG<storage::BitStorage, hashing::CanUnikmerShifter>;

    template class dBG<storage::SparseppSetStorage, hashing::FwdLemireShifter>;
    template class dBG<storage::SparseppSetStorage, hashing::CanLemireShifter>;
    template class dBG<storage::SparseppSetStorage, hashing::FwdUnikmerShifter>;
    template class dBG<storage::SparseppSetStorage, hashing::CanUnikmerShifter>;

    template class dBG<storage::ByteStorage, hashing::FwdLemireShifter>;
    template class dBG<storage::ByteStorage, hashing::CanLemireShifter>;
    template class dBG<storage::ByteStorage, hashing::FwdUnikmerShifter>;
    template class dBG<storage::ByteStorage, hashing::CanUnikmerShifter>;

    template class dBG<storage::NibbleStorage, hashing::FwdLemireShifter>;
    template class dBG<storage::NibbleStorage, hashing::CanLemireShifter>;
    template class dBG<storage::NibbleStorage, hashing::FwdUnikmerShifter>;
    template class dBG<storage::NibbleStorage, hashing::CanUnikmerShifter>;

    template class dBG<storage::QFStorage, hashing::FwdLemireShifter>;
    template class dBG<storage::QFStorage, hashing::CanLemireShifter>;
    template class dBG<storage::QFStorage, hashing::FwdUnikmerShifter>;
    template class dBG<storage::QFStorage, hashing::CanUnikmerShifter>;


    template class hashing::KmerIterator<dBG<storage::BitStorage, hashing::FwdLemireShifter>>;
    template class hashing::KmerIterator<dBG<storage::BitStorage, hashing::CanLemireShifter>>;
    template class hashing::KmerIterator<dBG<storage::BitStorage, hashing::FwdUnikmerShifter>>;
    template class hashing::KmerIterator<dBG<storage::BitStorage, hashing::CanUnikmerShifter>>;

    template class hashing::KmerIterator<dBG<storage::SparseppSetStorage, hashing::FwdLemireShifter>>;
    template class hashing::KmerIterator<dBG<storage::SparseppSetStorage, hashing::CanLemireShifter>>;
    template class hashing::KmerIterator<dBG<storage::SparseppSetStorage, hashing::FwdUnikmerShifter>>;
    template class hashing::KmerIterator<dBG<storage::SparseppSetStorage, hashing::CanUnikmerShifter>>;

    template class hashing::KmerIterator<dBG<storage::ByteStorage, hashing::FwdLemireShifter>>;
    template class hashing::KmerIterator<dBG<storage::ByteStorage, hashing::CanLemireShifter>>;
    template class hashing::KmerIterator<dBG<storage::ByteStorage, hashing::FwdUnikmerShifter>>;
    template class hashing::KmerIterator<dBG<storage::ByteStorage, hashing::CanUnikmerShifter>>;

    template class hashing::KmerIterator<dBG<storage::NibbleStorage, hashing::FwdLemireShifter>>;
    template class hashing::KmerIterator<dBG<storage::NibbleStorage, hashing::CanLemireShifter>>;
    template class hashing::KmerIterator<dBG<storage::NibbleStorage, hashing::FwdUnikmerShifter>>;
    template class hashing::KmerIterator<dBG<storage::NibbleStorage, hashing::CanUnikmerShifter>>;

    template class hashing::KmerIterator<dBG<storage::QFStorage, hashing::FwdLemireShifter>>;
    template class hashing::KmerIterator<dBG<storage::QFStorage, hashing::CanLemireShifter>>;
    template class hashing::KmerIterator<dBG<storage::QFStorage, hashing::FwdUnikmerShifter>>;
    template class hashing::KmerIterator<dBG<storage::QFStorage, hashing::CanUnikmerShifter>>;
}

File: goetia/dbg.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# (c) Camille Scott, 2019
# File   : dbg.py
# License: MIT
# Author : Camille Scott <camille.scott.w@gmail.com>
# Date   : 14.10.2019

from goetia import libgoetia
from goetia.hashing import typenames as hasher_types
from goetia.storage import get_storage_args, process_storage_args


dBG = libgoetia.dBG
STATES = libgoetia.TraversalState

def Graph(storage_type, shifter_type):

    def build(K, storage_args, shifter_args):
        storage = storage_type.build(*storage_args)
        shifter = shifter_type.build(K, *shifter_args)
        graph_type = dBG[storage_type, shifter_type]

        return graph_type.build(storage, hasher)

    return build


def get_graph_args(parser):
    group = parser.add_argument_group('dBG')
    group.add_argument('-K', '--ksize', type=int, default=31)
    group.add_argument('-H', '--hasher', choices=[name for t, name in hasher_types],
                       default='FwdLemireShifter')

    get_storage_args(parser)

    return group


def process_graph_args(args):
    process_storage_args(args)

    args.hasher_t = getattr(libgoetia.hashing, args.hasher)
    args.graph_t = libgoetia.dBG[args.storage, args.hasher_t]


def print_dBG_args(args):
    print('* dBG will be order', args.ksize, file=sys.stderr)
    print('* dBG will have underlying storage of', args.storage, file=sys.stderr)
    print('*','*' * 10, '*', sep='\n', file=sys.stderr)
